src/Stratis.Bitcoin.Features.BlockStore/BlockStoreBehavior.cs

```diff
@@ -131,49 +141,180 @@ protected override void DetachCore()
-            if (sendCmpctPayload != null)
+            // Now we want to find the last common block between our chain and the block locator the peer sent us.
+            ChainedBlock chainTip = this.chain.Tip;
+            ChainedBlock forkPoint = null;
+
+            // Find last common block between our chain and the block locator the peer sent us.
+            while (forkPoint == null)
             {
-                Task res = this.ProcessSendCmpctPayload(node, sendCmpctPayload);
-                this.logger.LogTrace("(-)[SEND_CMPCT_PAYLOAD]");
-                return res;
+                forkPoint = this.chain.FindFork(getBlocksPayload.BlockLocators.Blocks);
+                if (forkPoint == null)
+                {
+                    this.logger.LogTrace("(-)[NO_FORK_POINT]");
+                    return;
+                }
+
+                // In case of reorg, we just try again, eventually we succeed.
+                if (chainTip.FindAncestorOrSelf(forkPoint) == null)
+                {
+                    chainTip = this.chain.Tip;
+                    forkPoint = null;
+                }
             }
 
-            var sendHeadersPayload = message.Message.Payload as SendHeadersPayload;
-            if (sendHeadersPayload != null)
-                this.PreferHeaders = true;
+            this.logger.LogTrace("Fork point is '{0}'.", forkPoint);
+
+            // If block store is lower than the fork point, or it is on different chain, we don't have anything to contribute to this peer at this point.
+            if (blockStoreTip.FindAncestorOrSelf(forkPoint) == null)
+            {
+                this.logger.LogTrace("(-)[FORK_OUTSIDE_STORE]");
+                return;
+            }
+
+            // Now we compile a list of blocks we want to send to the peer as inventory vectors.
+            // This is needed as we want to traverse the chain in forward direction.
+            int maxHeight = Math.Min(blockStoreTip.Height, forkPoint.Height + InvPayload.MaxGetBlocksInventorySize);
+            ChainedBlock lastBlock = blockStoreTip.GetAncestor(maxHeight);
+            int headersCount = maxHeight - forkPoint.Height;
+
+            this.logger.LogTrace("Last block to announce is '{0}', number of blocks to announce is {1}.", lastBlock, headersCount);
+
+            ChainedBlock[] headersToAnnounce = new ChainedBlock[headersCount];
+            for (int i = headersCount - 1; i >= 0; i--)
+            {
+                headersToAnnounce[i] = lastBlock;
+                lastBlock = lastBlock.Previous;
+            }
+
+            // Now we compile inventory payload and we also consider hash stop given by the peer
+            // and we also check if the blocks to be announced are really present in the store.
+            bool sendContinuation = true;
+            ChainedBlock lastAddedChainedBlock = null;
+            var inv = new InvPayload();
+            for (int i = 0; i < headersToAnnounce.Length; i++)
```

Comment:
Looks like you could use a foreach here instead of a for

Reply:
I could, but it felt more aesthetic to use for when we have another one just above :-)
Do we want to prefer foreaches or is it OK to leave this one as is?

Reply:
I'm ok with using a for loop (I came from a C++ background 😁) but I think generally C#ers prefer to use foreach when iterating collections unless you are using the index for something else. At any rate just a suggestion I don't think this PR should be held up for this minor detail.